server/gui/main.py

```diff
@@ -7,33 +8,60 @@
 from PySide2.QtCore import *
 from PySide2.QtWidgets import *
 
-from server.app.app import Server
 from server.gui.browser import CefWidget, CefApplication
-from server.gui.workers import DataLoadWorker, ServerRunWorker
-from server.gui.utils import WINDOWS, LINUX, MAC, FileLoadSignals
+from server.gui.workers import Worker, SiteReadyWorker
+from server.gui.utils import WINDOWS, LINUX, MAC, FileLoadSignals, Emitter, WorkerSignals
 from server.utils.constants import MODES
+from server.utils.utils import find_available_port
 
 
 # Configuration
 # TODO remember this or calculate it?
 WIDTH = 1024
 HEIGHT = 768
+GUI_PORT = find_available_port("localhost")
 
-# noinspection PyUnresolvedReferences
 class MainWindow(QMainWindow):
     def __init__(self):
         super(MainWindow, self).__init__(None)
         self.cef_widget = None
         self.data_widget = None
-        self.server = Server()
-        self.server.create_app()
-        self.runServer()
+        self.parent_conn, self.child_conn = Pipe()
+        self.load_emitter = Emitter(self.parent_conn, WorkerSignals)
+        self.load_emitter.signals.error.connect(self.restartOnError)
+        self.emitter_thread = threading.Thread(target=self.load_emitter.run, daemon=True)
+        self.emitter_thread.start()
+        self.worker = None
+        self.url = f"http://localhost:{GUI_PORT}/"
         self.setWindowTitle("cellxgene")
 
         # Strong focus - accepts focus by tab & click
         self.setFocusPolicy(Qt.StrongFocus)
         self.setupLayout()
-        self.setupMenu()
+        # self.setupMenu()
+
+    def restartOnError(self):
+        if self.worker:
+            self.terminateWorker()
+        self.parent_conn.close()
+        # close emitter on error/finished
+        self.parent_conn, self.child_conn = Pipe()
+        self.load_emitter = Emitter(self.parent_conn, WorkerSignals)
+        self.emitter_thread = threading.Thread(target=self.load_emitter.run, daemon=True)
+        self.emitter_thread.start()
+        # send to load with error message?
+
+    def terminateWorker(self):
+        if not self.worker:
+            return
+        try:
+            self.worker.close()
```

Comment:
Left comment on this in the other PR.  TL;DR - not sure why you can't just call `terminate()` in all cases.

Reply:
I checked, and you're right. Terminate is always going to be called.